src/processors/default/ipfsMediaUploader.ts

```diff
@@ -0,0 +1,51 @@
+
+import { urlSource } from 'ipfs-http-client';
+import _ from 'lodash'
+
+import { Table } from '../../db/db';
+import { artworkNotOnIpfs, audioNotOnIpfs } from '../../triggers/ipfs';
+import { Clients, Processor } from '../../types/processor';
+import { ProcessedTrack } from '../../types/track';
+
+function processorFunction(sourceField: 'lossyAudioURL' | 'lossyArtworkURL', replaceField: 'lossyAudioIPFSHash' | 'lossyArtworkIPFSHash') {
+  return async (newTriggerItems: any, clients: Clients) => {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```

Comment:
comment required because TS doesn't recognise the optional field `processedTrack.lossyAudioURL`. @musnit do you know how to fix this?

Reply:
i think you can use something like:

```
 if ('lossyAudioURL' in processedTrack) {
 ...
} 
```
to check for the existence of a field on an object and then handle an error if that field doesn't exist as expected

Reply:
I tried this but I get the error:

Element implicitly has an 'any' type because expression of type '"lossyAudioURL" | "lossyArtworkURL"' can't be used to index type 'ProcessedTrack'.
  Property 'lossyAudioURL' does not exist on type 'ProcessedTrack'


I notice we don't access processedTrack.lossyAudioURL or processedTrack.lossyAudioIPFSHash anywhere else in the codebase so this seems like a problem with how TS is managing those optional fields?

Reply:
yeh so basically with TS you cannot access optional fields unless you know that they are there. using object.field! doesn't work because ! just tells TS that this field is not undefined, it does not tell TS that the field actually exists (existence is kinda a separate thing to being undefined)

so for example

a= {}
b = {name:undefined}

these are actually 2 different things, i believe both in JS and TS but definitely in TS so a.name! won't work because TS knows that name is not a field on this object.

the if (field in object) check should allow you to get typescript to confirm whether or not a field actually is there, so for example

```
 if ('name' in a) {
   a.name!
} 
```

this should work afaik, as long as you only access the field within the conditional.

so in our case, did you try wrap all the logic/accesses you want inside a 

```
 if ('lossyAudioURL' in processedTrack) {
 ...
} 
```

block?

(also, i'm not sure about whether `if(sourceField in processedTrack)` works or only a hardcoded `if('lossyAudioURL' in processedTrack)` works, hopefully both?)

and yeh this is one of the weirded edge cases of typescript so if the above suggestion doesn't work then sure let's just leave it as you've already done.

Reply:
yeah unfortunately I didn't have any success with that technique so just leaving the comments in for now